src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs

```diff
@@ -0,0 +1,157 @@
+
+            return status;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        public override ODataAPIResponseStatus TryCreate(IDictionary<string, object> keyValues, out TStructuralType createdObject, out string errorMessage)
+        {
+            createdObject = default(TStructuralType);
+            errorMessage = string.Empty;
+
+            try
+            {
+                if(originalList != null)
+                {
+                    originalList = new List<TStructuralType>();
+                }
+
+                createdObject = Activator.CreateInstance(_clrType) as TStructuralType;
+                originalList.Add(createdObject);
+
+                return ODataAPIResponseStatus.Success;
+            }
+            catch (Exception ex)
+            {
+                errorMessage = ex.Message;
+
+                return ODataAPIResponseStatus.Failure;
+            }
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        public override ODataAPIResponseStatus TryDelete(IDictionary<string, object> keyValues, out string errorMessage)
+        {
+            errorMessage = string.Empty;
+
+            try
+            {
+                TStructuralType originalObject = GetFilteredItem(keyValues);
+                originalList.Remove(originalObject);
+
+                return ODataAPIResponseStatus.Success;
+            }
+            catch (Exception ex)
+            {
+                errorMessage = ex.Message;
+
+                return ODataAPIResponseStatus.Failure;
+            }
+        }
+
+        public override IODataAPIHandler GetNestedHandler(TStructuralType parent, string navigationPropertyName)
+        {
+            foreach (PropertyInfo property in _clrType.GetProperties())
+            {
+                if (property.Name == navigationPropertyName)
+                {
+                    Type type = typeof(DefaultODataAPIHandler<>).MakeGenericType(property.PropertyType.GetGenericArguments()[0]);
+
+                    return Activator.CreateInstance(type, property.GetValue(parent)) as IODataAPIHandler;
+                }
+            }
+
+            return null;
+        }
+
+
+        private TStructuralType GetFilteredItem(IDictionary<string, object> keyValues)
+        {
+            //This logic is for filtering the object based on the set of keys,
+            //There will only be very few key elements usually, mostly 1, so performance wont be impacted.
+
+            if(originalList == null || originalList.Count == 0)
+            {
+                return default(TStructuralType);
+            }
+
+            Dictionary<string, PropertyInfo> propertyInfos = new Dictionary<string, PropertyInfo>();
+
+            foreach (string key in keyValues.Keys)
+            {
+                propertyInfos.Add(key, _clrType.GetProperty(key));
```

Comment:
Is there any validation done to ensure that the `keyValues` dictionary contains only key-value pairs where key MUST correspond to a property in the `_clrType`? If no, `GetPropety` can return null and you may need to confirm that the statements that follow will handle that just fine

Reply:
No validation for now.
Complex scenarios and validations should be handled in a custom `ODataAPIHandler`